core/src/components/range/test/range-events.e2e.ts

```diff
@@ -0,0 +1,132 @@
+  });
+
+  test.describe(' ionChange', () => {
+    test('should not emit if the value is set programmatically', async ({ page }) => {
+      await page.setContent(`<ion-range></ion-range>`);
+
+      const range = page.locator('ion-range');
+      const ionChangeSpy = await page.spyOnEvent('ionChange');
+
+      await range.evaluate((el: HTMLIonRangeElement) => {
+        el.value = 50;
+      });
+
+      await page.waitForChanges();
+
+      expect(ionChangeSpy).toHaveReceivedEventTimes(0);
+
+      // Update the value again to make sure it doesn't emit a second time
+      await range.evaluate((el: HTMLIonRangeElement) => {
+        el.value = 60;
+      });
+
+      await page.waitForChanges();
+
+      expect(ionChangeSpy).toHaveReceivedEventTimes(0);
+    });
+
+    test('should emit when the knob is released', async ({ page }) => {
+      await page.setContent(`<ion-range></ion-range>`);
+
+      const rangeHandle = page.locator('ion-range .range-knob-handle');
+      const ionChangeSpy = await page.spyOnEvent('ionChange');
+
+      const boundingBox = await rangeHandle.boundingBox();
+
+      await rangeHandle.hover();
+      await page.mouse.down();
+      await page.mouse.move(boundingBox!.x + 100, boundingBox!.y);
+
+      await page.mouse.up();
+
+      await ionChangeSpy.next();
+
+      expect(ionChangeSpy).toHaveReceivedEvent();
+      expect(ionChangeSpy).toHaveReceivedEventTimes(1);
+    });
+
+    test('should emit when the knob is moved with the keyboard', async ({ page }) => {
+      await page.setContent(`<ion-range value="50"></ion-range>`);
+
+      const rangeHandle = page.locator('ion-range .range-knob-handle');
+      const ionChangeSpy = await page.spyOnEvent('ionChange');
+
+      await rangeHandle.click();
+
+      await page.keyboard.press('ArrowLeft');
+      await ionChangeSpy.next();
+
+      expect(ionChangeSpy).toHaveReceivedEventDetail({ value: 49 });
+
+      await page.keyboard.press('ArrowRight');
+      await ionChangeSpy.next();
+
+      expect(ionChangeSpy).toHaveReceivedEventDetail({ value: 50 });
+
+      await page.keyboard.press('ArrowUp');
+      await ionChangeSpy.next();
+
+      expect(ionChangeSpy).toHaveReceivedEventDetail({ value: 51 });
+
+      await page.keyboard.press('ArrowDown');
+      await ionChangeSpy.next();
+
+      expect(ionChangeSpy).toHaveReceivedEventDetail({ value: 50 });
+    });
+  });
+
+  test.describe('ionInput', () => {
+    test('should emit when the knob is dragged', async ({ page }) => {
+      await page.setContent(`<ion-range></ion-range>`);
+
+      const rangeHandle = page.locator('ion-range .range-knob-handle');
+      const ionInputSpy = await page.spyOnEvent('ionInput');
+
+      const boundingBox = await rangeHandle.boundingBox();
+
+      await rangeHandle.hover();
+      await page.mouse.down();
+      await page.mouse.move(boundingBox!.x + 100, boundingBox!.y);
+
+      await ionInputSpy.next();
+
+      expect(ionInputSpy).toHaveReceivedEvent();
```

Comment:
Do we want to check that `ionInput` was emitted multiple times throughout the drag? Unsure if `toHaveReceivedEventTimes` would be flaky or something.

Reply:
I opted against checking for a specific number of emissions with similar concerns of flaky tests. Since the mouse pointer is not released, this does cover the expected behavior of the `ionInput` event being emitted as the value is being updated (and before `ionChange` is emitted).